patchwork/context.py

```diff
@@ -113,6 +113,19 @@ def is_own_ancestor(self, db: Datastore) -> bool:
             context = context.parent
         return False
 
+    def is_blocking(self, db: Datastore) -> bool:
+        """Return whether anyone is waiting for an answer from this context."""
+        return db.has_promisees(
+                    db.dereference(self.workspace_link).answer_promise)
+
+    def can_fulfill_promise(self, db: Datastore, promise: Address) -> bool:
+        return promise == db.dereference(self.workspace_link).answer_promise
+
+    def can_advance_promise(self, db: Datastore, promise: Address) -> bool:
+        return self.can_fulfill_promise(db, promise) \
+               or (self.is_blocking(db)
+                   and self.parent.can_advance_promise(db, promise))
```

Comment:
How can I see that this logic covers all cases? Maybe I can't directly give the promised answer, and maybe none of my ancestors can either, but might I not still be blocking some other context that can answer or whose ancestors can? After all, with lazy evaluation, my pointer can be passed to all kinds of places before it's evaluated.

Reply:
Good point. I didn't think through the case where `choose_context_to_advance_promise` is applied to a promise P other than the root answer promise or links of the root answer. So `can_advance_promise` shouldn't check whether a parent can advance P, but whether one of self's promisees can advance P.

Also, I didn't take into account the case when a context is trying to unlock an unfulfilled workspace promise.

Back to the punch cards…

Reply:
```
What is your root question?
> Root?
Question: [$1: Root?]
Scratchpad: [$2: ]
Subquestions:


> ask Sub3?
Question: [$1: Root?]
Scratchpad: [$2: ]
Subquestions:
1.
  [$q1: Sub3?]
  $a1
  $w1

> ask Sub2 ($a1)?
Question: [$1: Root?]
Scratchpad: [$2: ]
Subquestions:
1.
  [$q1: Sub3?]
  $a1
  $w1
2.
  [$q2: Sub2 ($a1)?]
  $a2
  $w2

> reply $a2
Question: [$1: Sub2 ($3)?]
Scratchpad: [$2: ]
Subquestions:


> unlock $3
Encountered an error with your command: 
Question: [$1: Sub2 ($3)?]
Scratchpad: [$2: ]
Subquestions:


> Traceback (most recent call last):
  File "/home/erle/repos/patchwork/patchwork/interface.py", line 40, in _do
    result = self.session.act(action)
  File "/home/erle/repos/patchwork/patchwork/scheduling.py", line 255, in act
    self.current_context = resulting_context or self._choose_next_context()
  File "/home/erle/repos/patchwork/patchwork/scheduling.py", line 227, in _choose_next_context
    resulting_context = self.sched.choose_context_to_advance_promise(self.promise_to_advance)
  File "/home/erle/repos/patchwork/patchwork/scheduling.py", line 182, in choose_context_to_advance_promise
    choice = next(c for c in self.pending_contexts
  File "/home/erle/repos/patchwork/patchwork/scheduling.py", line 183, in <genexpr>
    if c.can_advance_promise(self.db, promise))
  File "/home/erle/repos/patchwork/patchwork/context.py", line 127, in can_advance_promise
    and self.parent.can_advance_promise(db, promise))
  File "/home/erle/repos/patchwork/patchwork/context.py", line 126, in can_advance_promise
    or (self.is_blocking(db)
  File "/home/erle/repos/patchwork/patchwork/context.py", line 119, in is_blocking
    db.dereference(self.workspace_link).answer_promise)
  File "/home/erle/repos/patchwork/patchwork/datastore.py", line 54, in has_promisees
    return bool(self.promises[address])
KeyError: Address(9ae6e745-702a-11e8-8cf8-9c4e363c78a4)
```

Reply:
```python
    def get_promisee_contexts(self, db: Datastore) -> List[Context]:
        return [Context(args[0], db)
                for args in db.get_promisees(db.dereference(self.workspace_link)
                                             .answer_promise)]

    def can_advance_promise(self, db: Datastore, promise: Address) -> bool:
        return self.can_fulfill_promise(db, promise) or \
               (self.is_blocking(db)
                and any(pc.can_advance_promise(db, promise)
                        for pc in self.get_promisee_contexts(db)))
```

This should avoid the non-ancestors issue. I still have to scrutinize it, though, because I'm probably overlooking something again. Also I haven't figured out the workspace case yet.